### PR TITLE
Manage package and service

### PR DESCRIPTION
**Pull request by jerith**

### CODE REVIEW

File: manifests/repo.pp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+#   Repo source. Valid sources are:
+#   * `gluster/glusterfs-3.7`
+#
+class gluster::repo(
+  $manage = true,
+  $source = 'gluster/glusterfs-3.7',
+) {
+  if $manage {
+    case $::osfamily {
+      'Debian': {
```
Comment (JayH5) [inline manifests/repo.pp]: Can we filter on Ubuntu somehow? I know we're not supporting Debian but the PPA won't work on Debian.

Comment (jerith) [inline manifests/repo.pp]: We can, and I plan to. I just need to find a clean way to do it.

File: manifests/init.pp [hunk 2]
```diff
@@ -1,5 +1,71 @@
...
+  $package_manage = $gluster::params::package_manage,
+  $package_ensure = $gluster::params::package_ensure,
+  $package_name   = $gluster::params::package_name,
+  $service_manage = $gluster::params::service_manage,
+  $service_ensure = $gluster::params::service_ensure,
+  $service_name   = $gluster::params::service_name,
+) inherits gluster::params {
+
+  validate_bool($repo_manage)
+  validate_bool($package_manage)
```
Comment (JayH5) [inline manifests/init.pp]: Should validate `service_manage`

**Comment by JayH5** [comment]
The reason I did the `manage` checks inside the "sub-classes" in the Marathon module is because it makes it a lot easier to tie up the dependencies between the different classes.

At the moment I don't see anything that would prevent the service from being created before the package is installed.

What is the use case for making `gluster::repo`, `gluster::install`, and `gluster::service` available to be used even when `gluster` is included? A lot of modules I've seen actually go further and make it impossible to use the sub-classes outside of the main class, whether the main class is included or not.

File: manifests/init.pp [hunk 3]
```diff
@@ -1,5 +1,71 @@
...
+  # If $service_manage is false, we still want to allow `gluster::service` to
+  # be used outside this class.
+  if $service_manage {
+    class { 'gluster::service':
+      ensure       => $service_ensure,
+      manage       => $service_manage,
+      service_name => $service_name,
+    }
+  }
+
```
Comment (JayH5) [inline manifests/init.pp]: Generally want to anchor or contain the classes included here.

File: manifests/init.pp (hunk 2 above)
Comment (jerith) [inline manifests/init.pp]: Good catch, thanks.

**Comment by jerith** [comment]
I'm not really sure, I didn't actually think that deeply about it. Thanks for pointing that out. :-)

File: manifests/service.pp [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  $ensure       = $gluster::params::service_ensure,
+  $service_name = $gluster::params::service_name,
+) inherits gluster::params {
+  if $manage {
+    service { $service_name:
+      ensure => $ensure,
+    }
+    if $::osfamily == 'Debian' {
+      include apt
+      Class['apt::update'] -> Service[$service_name]
```
Comment (JayH5) [inline manifests/service.pp]: Why do we need this?

File: manifests/install.pp [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  $ensure       = $gluster::params::package_ensure,
+  $package_name = $gluster::params::package_name,
+) inherits gluster::params {
+  if $manage {
+    package { $package_name:
+      ensure => $ensure,
+    }
+    if $::osfamily == 'Debian' {
+      include apt
+      Class['apt::update'] -> Package[$package_name]
```
Comment (JayH5) [inline manifests/install.pp]: If `Class['gluster::repo'] -> Class['gluster::install']` then this shouldn't be necessary. (Thanks to the `contain 'apt::update'` in `gluster::repo`)

File: manifests/service.pp (hunk 4 above)
Comment (jerith) [inline manifests/service.pp]: Oops, copy pasta.

File: manifests/install.pp (hunk 5 above)
Comment (jerith) [inline manifests/install.pp]: Good point, thanks.

**Comment by JayH5** [comment]
OK, one last thing. We got this PR today: https://github.com/praekeltfoundation/puppet-marathon/pull/15

May be worthwhile adding a package ~> service notification? I'm not sure if it's always necessary..

**Comment by JayH5** [comment]
:+1:

**Comment by JayH5** [comment]
:+1:

**Pull request merged**